### PR TITLE
Tinkerpop 968:Add first class support for an optional traversal

### PR DESCRIPTION
**Pull request by pietermartin**
https://issues.apache.org/jira/browse/TINKERPOP-968
Added `optional(Traversal)` to `GraphTraversal`. Optional semantics is implemented via the `choose` step.
Added `OptionalTest.java`
Added `Optional`-step documentation.

### CODE REVIEW

File: docs/src/reference/the-traversal.asciidoc [hunk 1]
```diff
@@ -1127,6 +1127,31 @@ g.V().repeat(both()).times(3).values('age').min()
 IMPORTANT: `min(local)` determines the min of the current, local object (not the objects in the traversal stream).
 This works for `Collection` and `Number`-type objects. For any other object, a min of `Double.NaN` is returned.
 
+[[optional-step]]
+Optional Step
+~~~~~~~~~~~~~
+
+The `optional()`-step returns the result of the specified traversal if it yields a result else it returns the calling
```
Comment (okram) [inline docs/src/reference/the-traversal.asciidoc]: please say that it is a `(*map*)-step` ... see the convention I use for other steps in the traversal.asciidoc.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/branch/OptionalTest.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        public Traversal<Vertex, Vertex> get_g_VX2X_optionalXinXknowsXX(Object v2Id) {
+            return this.g.V(v2Id).optional(in("knows"));
+        }
+
+        @Override
+        public Traversal<Vertex, Path> get_g_V_hasLabelXpersonX_optionalXoutXknowsX_optionalXoutXcreatedXXX_path() {
+            return this.g.V().hasLabel("person").optional(out("knows").optional(out("created"))).path();
+        }
+
+        @Override
```
Comment (okram) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/branch/OptionalTest.java]: Formatting is a big fuggly here. Why not just make it a one-liner?

**Comment by okram** [comment]
@pietermartin -- From just a code review, everything looks really clean and consistent with the conventions thus far. You added to `traversal.asciidoc`, you created a `GroovyOptionalTest` and an `OptionalTest`. Your test cases good. Finally, you were smart about adding `__.optional()`. Did my `GraphTraversalTest` open your eyes :) ?. 

Please make sure the docs build. Better yet, `publish-docs.sh` after your `process-docs.sh` and then we can all see the new section online. Finally, please do a full integration test. If you can paste your `BUILD SUCCESSFUL` console output of your integration test run, that would be great.  After that, I will vote.

Thank you Pieter.

**Comment by spmallette** [comment]
I'm not so sure that he has rights to do a publish-doc.she. I'm still a bit confused as to who else has access and under what circumstances they get access.

**Comment by pietermartin** [comment]
@okram Yes your GraphTraversalTest showed me the way.

mvn clean install  SUCCESS
mvn clean install -DskipIntegrationTests=false     

{code}
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Apache TinkerPop ................................... SUCCESS [  4.726 s]
[INFO] Apache TinkerPop :: Gremlin Shaded ................. SUCCESS [  1.531 s]
[INFO] Apache TinkerPop :: Gremlin Core ................... SUCCESS [ 26.490 s]
[INFO] Apache TinkerPop :: Gremlin Test ................... SUCCESS [  7.425 s]
[INFO] Apache TinkerPop :: Gremlin Groovy ................. SUCCESS [ 32.440 s]
[INFO] Apache TinkerPop :: Gremlin Groovy Test ............ SUCCESS [  3.856 s]
[INFO] Apache TinkerPop :: TinkerGraph Gremlin ............ SUCCESS [02:08 min]
[INFO] Apache TinkerPop :: Hadoop Gremlin ................. SUCCESS [03:46 min]
[INFO] Apache TinkerPop :: Spark Gremlin .................. SUCCESS [04:28 min]
[INFO] Apache TinkerPop :: Giraph Gremlin ................. SUCCESS [  02:29 h]
[INFO] Apache TinkerPop :: Neo4j Gremlin .................. SUCCESS [  1.696 s]
[INFO] Apache TinkerPop :: Gremlin Driver ................. SUCCESS [  5.887 s]
[INFO] Apache TinkerPop :: Gremlin Server ................. SUCCESS [11:38 min]
[INFO] Apache TinkerPop :: Gremlin Console ................ SUCCESS [01:11 min]
[INFO] Apache TinkerPop :: Gremlin Archetype .............. SUCCESS [  0.061 s]
[INFO] Apache TinkerPop :: Archetype - TinkerGraph ........ SUCCESS [  2.853 s]
[INFO] Apache TinkerPop :: Archetype - Server ............. SUCCESS [  8.433 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:54 h
[INFO] Finished at: 2016-03-29T19:42:09+02:00
[INFO] Final Memory: 83M/394M
[INFO] ------------------------------------------------------------------------
{code}

**Note** 2 assertions were commented out to make the integration test suite pass.
``GremlinServerAuthIntegrateTest.shouldFailAuthenticateWithPlainTextNoCredentials``
``GremlinServerAuthOldIntegrateTest.shouldFailAuthenticateWithPlainTextNoCredentials``
The correct exception is thrown but the assertion tests for the message which is different on my system for some as yet unknown reason.

bin/process-docs.sh
Checked the optional gremlin output, looks good.

vote +1

**Comment by okram** [comment]
Can you also check Neo4j? This is the "full integration" test command.

```
mvn clean install -DskipIntegrationTests=false -DincludeNeo4j
```

Given that you already did the `-DskipIntegrationTests`, just do:

```
mvn clean install -DincludeNeo4j
```

Finally, please talk with @spmallette about the `GremlinServerAuth` issue.

**Comment by spmallette** [comment]
We did discuss that. I have no idea why those won't work in his environment. I guess I will build this branch and see if there's any problems with Gremlin Server as a result of the change.

**Comment by pietermartin** [comment]
``mvn clean install -DincludeNeo4j``

{code}
[INFO] Reactor Summary:
[INFO] 
[INFO] Apache TinkerPop ................................... SUCCESS [  2.872 s]
[INFO] Apache TinkerPop :: Gremlin Shaded ................. SUCCESS [  1.384 s]
[INFO] Apache TinkerPop :: Gremlin Core ................... SUCCESS [ 23.977 s]
[INFO] Apache TinkerPop :: Gremlin Test ................... SUCCESS [  6.627 s]
[INFO] Apache TinkerPop :: Gremlin Groovy ................. SUCCESS [ 32.346 s]
[INFO] Apache TinkerPop :: Gremlin Groovy Test ............ SUCCESS [  3.490 s]
[INFO] Apache TinkerPop :: TinkerGraph Gremlin ............ SUCCESS [01:09 min]
[INFO] Apache TinkerPop :: Gremlin Benchmark .............. SUCCESS [  2.553 s]
[INFO] Apache TinkerPop :: Hadoop Gremlin ................. SUCCESS [02:36 min]
[INFO] Apache TinkerPop :: Spark Gremlin .................. SUCCESS [03:24 min]
[INFO] Apache TinkerPop :: Giraph Gremlin ................. SUCCESS [  3.875 s]
[INFO] Apache TinkerPop :: Neo4j Gremlin .................. SUCCESS [47:11 min]
[INFO] Apache TinkerPop :: Gremlin Driver ................. SUCCESS [  6.096 s]
[INFO] Apache TinkerPop :: Gremlin Server ................. SUCCESS [ 32.213 s]
[INFO] Apache TinkerPop :: Gremlin Console ................ SUCCESS [  9.870 s]
[INFO] Apache TinkerPop :: Gremlin Archetype .............. SUCCESS [  0.055 s]
[INFO] Apache TinkerPop :: Archetype - TinkerGraph ........ SUCCESS [  2.950 s]
[INFO] Apache TinkerPop :: Archetype - Server ............. SUCCESS [  8.241 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 56:39 min
[INFO] Finished at: 2016-03-30T14:39:07+02:00
[INFO] Final Memory: 96M/441M
[INFO] ------------------------------------------------------------------------
{code}

**Comment by okram** [comment]
VOTE +1.

**Comment by twilmes** [comment]
Code looks and good and it worked as expected when I played with it in the shell.  Nice addition.

VOTE: +1

**Comment by spmallette** [comment]
looks like enough votes to merge if @pietermartin votes himself. i'd still like to hear if anyone else can run gremlin server integration tests successfully against this before merge.

**Comment by pietermartin** [comment]
Just a ``mvn clean install -pl gremlin-server -DskipIntegrationTests=false`` will be good enough I'd say as its only there that the assertions were commented out.

**Comment by twilmes** [comment]
Just finished running `mvn verify -DskipIntegrationTests=false -pl gremlin-server` without issue.

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13:17 min
[INFO] Finished at: 2016-03-30T14:33:01-05:00
[INFO] Final Memory: 33M/395M
[INFO] ------------------------------------------------------------------------
```

**Comment by pietermartin** [comment]
Great, tx
VOTE: +1

**Comment by spmallette** [comment]
hmm - so it seems that the issue is with your environment @pietermartin   - weird. good to go for me.

**Pull request merged**